provider/awssd/aws_sd.go

```diff
@@ -264,6 +316,220 @@ func (p *AWSSDProvider) updatesToCreates(changes *plan.Changes) (creates []*endp
+					if err != nil {
+						return nil, nil, err
+					}
+					if _, ok := changesByHost[host]; !ok {
+						changesByHost[host] = &relevantChanges{}
+					}
+					chByHost := changesByHost[host]
+					if len(chByHost.SRVCre) == 0 {
+						chByHost.SRVCre = append(chByHost.SRVCre, ch)
+					}
+				}
+				// add a Create record type to the changesByHost. Add max 1 because only 1 A is expected for 1 host
+				if ch.RecordType == endpoint.RecordTypeA {
+					host := ch.DNSName
+					if _, ok := changesByHost[host]; !ok {
+						changesByHost[host] = &relevantChanges{}
+					}
+					chByHost := changesByHost[host]
+					if len(chByHost.ACre) == 0 {
+						chByHost.ACre = append(chByHost.ACre, ch)
+					} else {
+						log.Warnf("Skipping record %s because only 1 create change is expected for A record. This is unexpected.", ch.String())
+					}
+				}
+			}
+		}
+		if okNsDel {
+			for _, ch := range nsDeletes {
+				// add a Delete record type to the changesByHost. Add n records because n SRV deletes are expected
+				if ch.RecordType == endpoint.RecordTypeSRV {
+					_, host, _, _, err := p.srvHostTargetSplit(ch.Targets[0])
+					if err != nil {
+						return nil, nil, err
+					}
+					if _, ok := changesByHost[host]; !ok {
+						changesByHost[host] = &relevantChanges{}
+					}
+					chByHost := changesByHost[host]
+					chByHost.SRVDel = append(chByHost.SRVDel, ch)
+				}
+			}
+		}
+
+		// loop the map of relevant changes by host
+		for host, relCh := range changesByHost {
+			_, svcName := p.parseHostname(host)
+
+			// defome some relevant variables for handling the host-level changes
+			ASvc, ASvcExists := services[svcName]
+			var ASvcIsAAndSRV bool
+			if ASvcExists && p.isAAndSRVService(ASvc) {
+				ASvcIsAAndSRV = true
+			} else {
+				ASvcIsAAndSRV = false
+			}
+			var ACreExists bool
+			if len(relCh.ACre) == 1 {
+				ACreExists = true
+			} else {
+				ACreExists = false
+			}
+			var SRVCreExists bool
+			if len(relCh.SRVCre) == 1 {
+				SRVCreExists = true
+			} else {
+				SRVCreExists = false
+			}
+			var ACre *endpoint.Endpoint
+			if ACreExists {
+				ACre = relCh.ACre[0]
+			}
+			var SRVCre *endpoint.Endpoint
+			if SRVCreExists {
+				SRVCre = relCh.SRVCre[0]
+			}
+			SRVDeletes := relCh.SRVDel
+			var ASvcCurrentPrefSRV string
+			var ASvcCurrentPort string
+			if ASvcIsAAndSRV {
+				ASvcCurrentPrefSRV, _, err = p.srvDescrSplit(aws.StringValue(ASvc.Description))
+				if err != nil {
+					return nil, nil, err
+				}
+				_, _, _, _, ASvcCurrentPort, err = p.preferredSRVSplit(ASvcCurrentPrefSRV)
+				if err != nil {
+					return nil, nil, err
+				}
+			}
+			var SRVCrePrefSRV string
+			if SRVCreExists {
+				SRVCrePrefSRV, err = p.preferredSRVCombine(SRVCre.DNSName, SRVCre.RecordTTL, SRVCre.Targets[0])
+				if err != nil {
+					return nil, nil, err
+				}
+			}
+
+			// set the outputs of the main if-block
+			ASvcRemove := false
+			newPrefSRV := ""
+
+			//run the main if-block
```

Comment:
Those comments just show that the code need to be broken down to smaller pieces. The `if-else` branching became too complex and it's hard to follow.

Reply:
done, simplified and added detailed comments